### PR TITLE
HIVE-23891: UNION ALL and multiple task attempts can cause file duplication

### PR DESCRIPTION
**Pull request by dengzhhu653**
### What changes were proposed in this pull request?
For dynamic partition insert into a non-acid table, if the source query contains union all, a sub-directory with prefix "HIVE_UNION_SUBDIR_" will be created under the temp dynamic partition directory. When the job is finished, 
it searches all the files under the temp dynamic partition directory for removing the double-committed output files,
so if there is another sub-directory under the dynamic partition directory, it'll fail to remove the duplicate files resulting from multiple task attempts.

The PR tries to solve if there is "HIVE_UNION_SUBDIR_" sub-directory, then searches the output files under ${tmp_dynamic_partition_directory}/HIVE_UNION_SUBDIR_x for deduplication.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit test

### CODE REVIEW

**Comment by abstractdog** [comment]
> ### What changes were proposed in this pull request?
> ### Why are the changes needed?
> ### Does this PR introduce _any_ user-facing change?
> ### How was this patch tested?

@dengzhhu653: thanks for the patch, please describe the problem and solution here
I guess the problem is described correctly in jira, so a reader might be interested in the solution description here

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java [hunk 1]
```diff
@@ -162,6 +166,59 @@ public void testNonAcidDynamicPartitioning() throws Exception {
...
+    speculative.close(false);
+    Path[] paths = findFilesInBasePath();
+    List<Path> mondays = Arrays.stream(paths)
+        .filter(path -> path.getParent().toString().endsWith("partval=Monday/HIVE_UNION_SUBDIR_0"))
+        .collect(Collectors.toList());
+    Assert.assertTrue(mondays.size() == 2);
+    Set<String> fileNames = new HashSet<>();
+    fileNames.add(mondays.get(0).getName());
+    fileNames.add(mondays.get(1).getName());
+    Assert.assertTrue(fileNames.contains("000000_1") && fileNames.contains("000000_0"));
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:207]: please elaborate on this assert for useful assertion error messages like, like:
```
    Assert.assertTrue("000000_1 file is expected", fileNames.contains("000000_1"));
    Assert.assertTrue("000000_0 file is expected", fileNames.contains("000000_0"));

```

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java [hunk 2]
```diff
@@ -162,6 +166,59 @@ public void testNonAcidDynamicPartitioning() throws Exception {
...
+
+    op1.jobCloseOp(jobConf, true);
+    List<Path> resultFiles = new ArrayList<Path>();
+    recurseOnPath(basePath, basePath.getFileSystem(jc), resultFiles);
+    mondays = resultFiles.stream()
+        .filter(path -> path.getParent().toString().endsWith("partval=Monday/HIVE_UNION_SUBDIR_0"))
+        .collect(Collectors.toList());
+    Assert.assertTrue(mondays.size() == 1);
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:215]: please use Assert.assertEquals which will show the expected vs. actual values

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java [hunk 3]
```diff
@@ -162,6 +166,59 @@ public void testNonAcidDynamicPartitioning() throws Exception {
...
+    Assert.assertTrue(mondays.get(0).getName().equals("000000_1"));
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:216]: similarly,  Assert.assertEquals

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java (hunk 1 above)
Comment (dengzhhu653) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java]: done

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java [hunk 4]
```diff
@@ -183,37 +183,42 @@ public void testNonAcidRemoveDuplicate() throws Exception {
...
-    // speculative task also ends successfully
-    speculative.close(false);
+    // Assume op2 also ends successfully, this happens in different containers
+    op2.close(false);
     Path[] paths = findFilesInBasePath();
     List<Path> mondays = Arrays.stream(paths)
         .filter(path -> path.getParent().toString().endsWith("partval=Monday/HIVE_UNION_SUBDIR_0"))
         .collect(Collectors.toList());
-    Assert.assertTrue(mondays.size() == 2);
+    Assert.assertEquals("Two result files were created", 2, mondays.size());
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:203]: "Two result files were created" -> makes me think it's unexpected, instead we expect 2 files according to the parameters, can we have rather "2 results files are expected"

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java [hunk 5]
```diff
@@ -183,37 +183,42 @@ public void testNonAcidRemoveDuplicate() throws Exception {
...
     Set<String> fileNames = new HashSet<>();
     fileNames.add(mondays.get(0).getName());
     fileNames.add(mondays.get(1).getName());
-    Assert.assertTrue(fileNames.contains("000000_1") && fileNames.contains("000000_0"));
 
+    Assert.assertTrue("000000_1 file is expected", fileNames.contains("000000_1"));
+    Assert.assertTrue("000000_0 file is expected", fileNames.contains("000000_0"));
+
+    // This happens in HiveServer2 when the job is finished, the job will call
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:211]: are you sure it's HiveServer2? as jobCloseOp is part of the operator pipeline, they run as part of the execution in the tasks, am I missing something?

**Comment by abstractdog** [comment]
@dengzhhu653: minor comment regarding your latest code comment, otherwise, it's fine, I'm about to approve

File: ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java (hunk 5 above)
Comment (dengzhhu653) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/TestFileSinkOperator.java:211]: yeah, the `closeOp(boolean abort)` is called from task, but the `jobCloseOp` is called from the jobclient side(jobClose-> jobCloseOp):

https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java#L732-L748

**Review by abstractdog — APPROVED** [review]
+1

**Pull request merged**